In [ ]:
# mount your drive for reading data
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


## 1. Reading Data from Drive

In [ ]:
import pandas as pd
## Since the csv file has to be converted to data frame, we require pandas library

In [ ]:
Data=pd.read_csv("/content/drive/My Drive/Projects/IoT-Water_pumping_system-Predictive_Maintainance/Sensor Data/sensor.csv")
Data= Data.drop(['Unnamed: 0','sensor_15'], axis=1)
## Dropping sensor_15 column since it doesnt have any data in it

In [ ]:
Data.head(5)

timestamp  sensor_00  ...  sensor_51  machine_status
0  2018-04-01 00:00:00   2.465394  ...   201.3889          NORMAL
1  2018-04-01 00:01:00   2.465394  ...   201.3889          NORMAL
2  2018-04-01 00:02:00   2.444734  ...   203.7037          NORMAL
3  2018-04-01 00:03:00   2.460474  ...   203.1250          NORMAL
4  2018-04-01 00:04:00   2.445718  ...   201.3889          NORMAL

[5 rows x 53 columns]

In [ ]:
Data.shape
## There are 54 columns and 2,20,320 rows in the data frame
## Out of 54 columns 52 columns are sensors which will help us in predicting the target variable.Apart from which we have one timestamp and one machine status

(220320, 53)

## 2. Understanding Columns


In [ ]:
!pip install dtale

In [ ]:
import dtale
import dtale.app as dtale_app
dtale_app.USE_NGROK = True
dtale_app.USE_COLAB = True

In [ ]:
dtale.show(Data)

2020-11-10 19:03:26,395 - INFO     - NumExpr defaulting to 2 threads.


http://0ff045ab00ba.ngrok.io/dtale/main/1

In [ ]:
## Use this command dtale.get_instance(1).kill() to kill the Dtale instances
## USe this command dtale.instances() to get all running instances

# 3. Working on Time Column


In [ ]:
from datetime import *

In [ ]:
Data['timestamp']=pd.to_datetime(Data['timestamp'])
Data['Date']=Data['timestamp'].dt.date
Data['Time']=Data['timestamp'].dt.time

In [ ]:
Start_Date=Data.iloc[0,53]
print("Data started at ",Start_Date)

End_Date=Data.iloc[-1,53]
print("Data ended at ",End_Date)

Total_days=(End_Date-Start_Date).days
print("Total number of days in the data are ",Total_days)
print("Although the date difference shows 152. We need to consider 153 because we have observations till 23:59 on 2018-08-31")

Data started at  2018-04-01
Data ended at  2018-08-31
Total number of days in the data are  152
Although the date difference shows 152. We need to consider 153 because we have observations till 23:59 on 2018-08-31


In [ ]:
Data.head(5)

timestamp  sensor_00  ...        Date      Time
0 2018-04-01 00:00:00   2.465394  ...  2018-04-01  00:00:00
1 2018-04-01 00:01:00   2.465394  ...  2018-04-01  00:01:00
2 2018-04-01 00:02:00   2.444734  ...  2018-04-01  00:02:00
3 2018-04-01 00:03:00   2.460474  ...  2018-04-01  00:03:00
4 2018-04-01 00:04:00   2.445718  ...  2018-04-01  00:04:00

[5 rows x 55 columns]

In [ ]:
dtale.show(Data)

http://0ff045ab00ba.ngrok.io/dtale/main/2

In [ ]:
Data['Rows_fullNA'] = Data.apply(lambda x: x.count(), axis=1)

In [ ]:
full_miss_count=Data.loc[Data['Rows_fullNA'] == 4]
full_miss_count.shape

(11, 56)

## 4. Data Understanding

### TimeStamp feature observations

* Our data is from 01-04-2018 to 31-08-2018. 
* Each date has 1440 observations, meaning our data is collected on a 1-minute frequency (24*60=1440)
* Total number of days in records are 153. (153*24*60 = 220,320 records)

---

### Missing Value information

* 11 rows have all the sensor values missing [2018-04-27 17:48:00 to 2018-04-27 17:55:00, 2018-07-11 09:12:00 to 2018-07-11 09:14:00]
* 8 rows have 50% values missing 


---


### Skewness and Kurtosis
* 84% of the sens0r features (42 nos) are significantly skewed (<-1 or >1)
* Out of the skewed sensor data (12 nos =>28.5%) are positively skewed (Right skewed)
* 78% of sensor data (39 nos) is having substantial kurtosis. (<-3 or >3)
* Positive kurtosis can be interpreted as heavily peaked and negative kurtosis can be understood as too flat.
* With kurtosis we can say that most features are having sudden impulsive records
* Kurtosis greated than 3 means Leptokurtic. Also means heavy tails, that is extreme outliers


---
### Standard Deviation
* We have very high standard deviation for most of the sensor features
* This could be due to substantial leptokurtic data indicating impulses/outliers







[Results of Understanding of data](https://docs.google.com/spreadsheets/d/1O8oSpMMA3Rre-xdxKoTMplW4cVUgqjG_CS0y9DKNsrI/edit?usp=sharing)